Automattic/jetpack
# Search: Add a Gutenberg powered customization experience

EDIT: Please review #20282 instead of this PR.

#### Changes proposed in this Pull Request:
<img width="1792" alt="Screen Shot 2021-06-30 at 5 12 13 PM" src="https://user-images.githubusercontent.com/4044428/124042554-518ed800-d9c6-11eb-80c3-10cb06aebffd.png">

* Adds a fully functioning Jetpack Search customization experience powered by Gutenberg codenamed "Customberg" 😉.
  * Creates a new wp-admin page at `/wp-admin/admin.php?page=jetpack-search-customize`.
  * Serves a new React application that provides a Gutenberg-styled search customization experience.
  * Basic analytics instrumented via Tracks.
  * Fully responsive design.
* Updates Jetpack Search customization links in the Jetpack dashboard to point to Customberg on the following pages:
  * At-a-glance page.
  * My Plans page.
  * Performance settings page.
* Updates a `@wordpress` build package to ensure compatibility with `@wordpress/interface`, which Customberg relies on.
  * `@wordpress/dependency-extraction-webpack-plugin`
* Updates the search app build process, adds a new `watch-search` command.

---

#### Jetpack product discussion
See pcNPJE-7J-p2 and #20128.

#### Does this pull request change what data or activity we track or use?
Yes, we add additional Tracks analytics for the initial page view and when the save button is clicked.

---

#### Testing instructions:
Apply these changes to your site with a Jetpack Search subscription.

* Testing new links:
  * Visit the At-a-glance page (`/wp-admin/admin.php?page=jetpack#/dashboard`) and locate the "Customize your Search Experience" link inside the Search card. Ensure that it links to Customberg instead of the Customizer.
  * Visit the My Plans page (`/wp-admin/admin.php?page=jetpack#/my-plan`) and locate the "Customize Search" button inside the "Instant search and filtering" card. Ensure that it links to Customberg instead of the Customizer.
  * Visit the Performance settings page (`/wp-admin/admin.php?page=jetpack#/performance`). Ensure that you've enabled Instant Search and that there's a link with the text "Customize your Search experience." Ensure that it links to Customberg instead of the Customizer.

* Testing build and dependency updates:
  * Ensure that all existing JS tests pass as expected.
  * Ensure that `pnpm build` and `pnpm watch` work as expected.
  * Ensure that `pnpm build-search` and `pnpm watch-search` work as expected.

* Testing the new Search customization experience ("Customberg"):
  * Navigate to the new customization page at `/wp-admin/admin.php?page=jetpack-search-customize`.
  * Ensure that the page renders a Gutenberg layout with a header, a sidebar to the right, and the Jetpack Search application in the center/left.
  * Ensure that you can customize the search application via the controls in the options tab within the sidebar.
  * Ensure that the "Jetpack Search" tab renders a description of Customberg and includes a link to the Customizer.
  * Try resizing the page. Ensure that the page works as expected in all viewports ranging from mobile to desktop.
